### ISSUE TITLE
wgengine/magicsock: wgdev.Reconfig: unexpected endpoint for peer, expected the peer's public key

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#3088]
I can `tailscale ping` (disco ping) but not `tailscale ping --tsmp` a Linux 1.16.0 machine.

That Linux 1.16.0 machine's logs say:

```
Oct 14 13:28:16 tsbi tailscaled[7707]: wgdev.Reconfig: unexpected endpoint "84c0cfca85ac0aaf1086c76f7e7ff85cffb8ca98c5c87495da667ce3b09af306" for peer "[/Bifd]", expected the peer's public key
```

/cc @danderson @DentonGentry @crawshaw

**Comment by bradfitz** [comment]
I hacked up testcontrol to send a non-stop stream of changes (randomly adding/deleting/changing node/disco keys) and added some random sleeps in the client and promoted the error to panics, but so far I've been unable to reproduce it.

But a clue...

The error message says:

```
Oct 14 13:28:16 tsbi tailscaled[7707]: wgdev.Reconfig: unexpected endpoint "84c0cfca85ac0aaf1086c76f7e7ff85cffb8ca98c5c87495da667ce3b09af306" for peer "[/Bifd]", expected the peer's public key
```

That comes from:

```
case key.EqualString("endpoint"):
        // TODO: our key types are all over the place, and this
        // particular one can't parse a mem.RO or a []byte without
        // allocating. We don't reconfigure wireguard often though, so
        // this is okay.
        s := value.StringCopy()
        k, err := wgkey.ParseHex(s)
        if err != nil {
                return fmt.Errorf("invalid endpoint %q for peer %q, expected a hex public key", s, peer.PublicKey.ShortString())
        }
        if k != peer.PublicKey {
                return fmt.Errorf("unexpected endpoint %q for peer %q, expected the peer's public key", s, peer.PublicKey.ShortString())
        }
```

So how would they get out of sync?

The clue is `84c0cfca85ac0aaf1086c76f7e7ff85cffb8ca98c5c87495da667ce3b09af306` in short form is `[hMDPy]` which is my Mac desktop on my tailscale.com email account (nodeid:49969443346629937 100.90.243.7)

But `[/Bifd]` is the same physical machine & same Tailscale IPNExtension process, but logged into my gmail account (nodeid:23936249232029309 100.82.88.80).

Notably, they share the same disco key when switching accounts.

Further, the `tsbi` machine that got into that stuck state above is shared from my `gmail.com` account to my `tailscale.com` account.

The bug almost certainly related to one or both of those things.

**Comment by bradfitz** [comment]
Yeah, so because of sharing, `tsbi`'s network map will contain both my gmail and work nodes of my iMac, and if switch accounts while keeping it open, their discokeys will be the same, and then `tsbi` will have two nodes in its netmap with the same discokey.

IIRC we hit a variant of this in the past and I'd proposed filtering them away in control so that's never the case? Forget where that bug or discussion is.

**Comment by bradfitz** [comment]
Magicsock is almost all fixed up now to not assume that there's at most DiscoKey in a netmap.

A bit of cleanup in peerMap remains, though. Its `byDiscoKey map[tailcfg.DiscoKey]*peerInfo` field is currently still present (as of 569f70abfdfdf762122ab4a34c3683eee8b9ee97). Once that's gone, then this should be fixed.

**Comment by bradfitz** [comment]
BTW, a more common way to reproduce this without involving sharing is probably just admins having multiple accounts within a domain: like `services@` and `regular-user@` and switching back & forth.

**Comment by bradfitz** [comment]
There's still one notable problem with moving away from assuming a DiscoKey can map to exactly 1 node: disco pings don't contain their source NodeKey.

This is a problem because the wireguard-go interface we implement ([ReceiveFunc](https://pkg.go.dev/github.com/btctcn/wireguard-go/conn#ReceiveFunc])) needs to return an endpoint, and when reading a packet over UDP, all we have is the packet's src `ip:port`. So we need some lookup from `ip:port` to endpoint.

If we receive a `CallMeMaybe` (which arrives via DERP), then we know the source node key (we get it from DERP) as well as the disco key, so we can map from disco key => most recent node key.

But if the node trying to reach us is on the same LAN or we have an active portmapping, the disco ping will come directly, usually beating the `CallMeMaybe` message. That means we have a disco ping message from an `ip:port` and don't know which NodeKey it's for.

We can reply to the disco ping without knowing the node key, but once the wireguard data packet arrives, we need the NodeKey and don't know it.  At least for maybe 10-250 ms when the CallMeMaybe arrives. :-/

**Comment by crawshaw** [comment]
By the time we're modifying the disco protocol, I think we need to weigh it against the alternative: make the client regenerate discokeys at login, and prove to control it owns the key.

**Comment by bradfitz** [comment]
I don't see a nice upgrade path for existing clients involving modifying the protocol to prove things to control.

And adding some fields to a disco message is easy. We've done it in the past (adding `MyNumber` to `CallMeMaybe). Here: https://github.com/tailscale/tailscale/pull/3116

The point is really just to make local LAN connectivity work unambiguously with low latency without waiting for the CallMeMaybe from DERP (much higher latency than local LAN) to clarify the mapping.

**Comment by qbarrand** [comment]
Hello everyone,

I am affected by this on
```
1.16.2
  tailscale commit: f449111792f662c8552514cf3691c3c1bd7f38e4
  other commit: 3853a3b920c7ec33da2b8ba8454e9ca8b3772774
  go version: go1.17.1-ts097d1284f9
```
The node is a CentOS 7 machine. It exposes its public IPv4 IP as a /32 route in Tailscale.
After a while, this node (`A`) does not seem to be accepting connections from any other node connected to my tailnet (`B`).
Restarting `tailscaled` via `systemctl` makes everything work again.

Logs on `A`:
```
nov. 02 00:00:08 hostname-A tailscaled[8572]: wgengine: Reconfig: configuring userspace wireguard config (with 2/5 peers)
nov. 02 00:00:08 hostname-A tailscaled[8572]: wgcfg.Reconfig failed: unexpected endpoint "5c0f7f376cfd9e556dd7fd23f27384464112c8817991f54a1914720ded061
nov. 02 00:00:08 hostname-A tailscaled[8572]: wgdev.Reconfig: unexpected endpoint "5c0f7f376cfd9e556dd7fd23f27384464112c8817991f54a1914720ded06193f" fo
```

Logs on `B` when trying to connect to `A:443/tcp`:
```
nov. 02 00:04:56 hostname-B tailscaled[1292]: wgengine: Reconfig: configuring userspace wireguard config (with 1/5 peers)
nov. 02 00:04:56 hostname-B tailscaled[1292]: magicsock: disco: node [pM7wb] d:a7df8137c7e0d124 now using 85.209.49.115:41641
nov. 02 00:05:01 hostname-B tailscaled[1292]: open-conn-track: timeout opening (TCP tailscale-IP-B:43850 => A-exposed-IP:443) to node [pM7wb]; online=yes, lastRecv=never
nov. 02 00:05:04 hostname-B tailscaled[1292]: open-conn-track: timeout opening (TCP tailscale-IP-B:43850 => A-exposed-IP:443) to node [pM7wb]; online=yes, lastRecv=never
nov. 02 00:05:17 hostname-B tailscaled[1292]: open-conn-track: timeout opening (TCP tailscale-IP-B:43850 => A-exposed-IP:443) to node [pM7wb]; online=yes, lastRecv=never
```

Bug report from A is `BUG-efe8d6cf0b88cb8158d9ab44eb3a8f973c1bbfafc9689305b1bd90fda3bf35d3-20211101233649Z-7149189394f9febd`.

Hope that helps. Thanks a lot for such a great product.

**Comment by DentonGentry** [comment]
I think the `expected the peer's public key` issue was fixed in 1.16.1?

**Comment by josharian** [comment]
It was fixed in 41da7620af8508e415ac0b1e676f00d47ad5bbf4, which has not been cherry-picked to 1.16 (yet?).

**Comment by DentonGentry** [comment]
Ah, in that case: we released 1.18.0 earlier this afternoon, which does contain the fix.
Closing as fixed in 1.18.0
